lib/ui/text.dart

```diff
@@ -838,23 +838,42 @@ class TextBox {
 /// For example, when a text position exists at a line break, a single offset has
 /// two visual positions, one prior to the line break (at the end of the first
 /// line) and one after the line break (at the start of the second line). A text
-/// affinity disambiguates between those cases. (Something similar happens with
-/// between runs of bidirectional text.)
+/// affinity disambiguates between those cases.
+///
+/// This affects only line breaks caused by wrapping, not explicit newline
+/// characters. For newline characters, the position is fully specified by the
+/// offset alone, and there is no ambiguity.
+///
+/// TextAffinity also affects bidirectional text at the interface between
+/// LTR and RTL text. Consider the following string, where the lowercase letters
+/// will be displayed as LTR and the uppercase letters RTL: "helloHELLO".  The
+/// string would appear as "helloOLLEH". A text position of 6 would be
+/// ambiguous, since it could be represented as being to the right of the "o"
+/// or to the right of the "H".
 enum TextAffinity {
   /// The position has affinity for the upstream side of the text position.
   ///
-  /// For example, if the offset of the text position is a line break, the
```

Comment:
We really need to provide an intuitive definition of upstream here. If it can be explained in terms of characters being entered with a text field then perhaps upstream is towards the older characters, and downstream is towards the newer ones.

Since BIDI text is where this gets pretty complicated it might be worth stuffing a helloHELLO example in here and in the downstream doc.

It would also be worth reiterating what "text position" means. As I understand it, a text position indicates a location in between two characters in the internal representation of the text - not the displayed version.  So offset 5 in "helloHELLO" is in between the o and the H. But in the rendered version of the text, "helloOLLEH", offset 5 is either to the right of the "o" for upstream affinity, or to the right of the H for downstream. 

That's what I'm assuming is correct anwyay ...